src/Nethermind/Nethermind.Core/Eip2930/AccessList.cs

```diff
@@ -1,30 +1,109 @@
 // SPDX-FileCopyrightText: 2022 Demerzel Solutions Limited
 // SPDX-License-Identifier: LGPL-3.0-only
 
+using System;
 using System.Collections.Generic;
 using Nethermind.Int256;
 
-namespace Nethermind.Core.Eip2930
+namespace Nethermind.Core.Eip2930;
+
+public abstract record AccessListItem
 {
-    public class AccessList
+    public record Address(Core.Address Value) : AccessListItem;
+
+    public record StorageKey(UInt256 Value) : AccessListItem;
+}
+
+public class AccessList
+{
+    private readonly List<AccessListItem> _items;
+
+    private AccessList(List<AccessListItem> items)
+    {
+        _items = items;
+    }
+
+    public static AccessList Empty() => new AccessList(new List<AccessListItem>());
+
+    public IReadOnlyCollection<AccessListItem> Raw => _items;
+
+    public IReadOnlyDictionary<Address, IReadOnlySet<UInt256>> AsDictionary()
```

Comment:
This is used only in WarmUp, where I think going through a duplicate once in a while shouldn't be a problem. Hence this is completely unnecessary.

Reply:
Another thing is I think we can assume there is always an address before any indexes in the list.

Reply:
My proposition to change it to this method for places where it is needed:

```
    public IEnumerable<(Address, IEnumerable<UInt256>)> AsEnumerable()
    {
        IEnumerable<UInt256> GetStorageKeys(int i)
        {
            while (i < _items.Count && _items[i] is AccessListItem.StorageKey storageKey)
            {
                yield return storageKey.Value;
                i++;
            }
        }
        for (int i = 0; i < _items.Count; i++)
        {
            AccessListItem item = _items[i];
            switch (item)
            {
                case AccessListItem.Address address:
                {
                    yield return (address.Value, GetStorageKeys(i + 1));
                    break;
                }
            }
        }
    }
```

Reply:
If validation needed:

```
    public IEnumerable<(Address, IEnumerable<UInt256>)> AsEnumerable()
    {
        IEnumerable<UInt256> GetStorageKeys(int i)
        {
            while (i < _items.Count && _items[i] is AccessListItem.StorageKey storageKey)
            {
                yield return storageKey.Value;
                i++;
            }
        }

        bool orphanedIndex = true;
        for (int i = 0; i < _items.Count; i++)
        {
            AccessListItem item = _items[i];
            switch (item)
            {
                case AccessListItem.Address address:
                {
                    orphanedIndex = false;
                    yield return (address.Value, GetStorageKeys(i + 1));
                    break;
                }
                case AccessListItem.StorageKey when orphanedIndex:
                {
                    throw new ArgumentException("Invalid access list, index without address");
                }
            }
        }
    }
```

Reply:
Can be `InvalidDataException` too if preffered

Reply:
The idea behind `AsDictionary` was to preserve the old API as much as possible, but indeed, it's only used in the `WarmUp` method, so we could safely remove that if we don't care about warming up duplicates.

Reply:
As for validation, it would not be required since the only (public/safe) way to build an `AccessList` through the `AccessList.Builder`, which already handles validation.